### PR TITLE
libc: time.h: add test groups for gmtime, mktime, strftime

### PR DESCRIPTION
**Pull request by jmaksymowicz**
Moved tests for time functions (gmtime, mktime and strftime) that existed previously and added them to libc automatic testing.

## Description

Tests for the mentioned time functions existed, but were never compiled or checked. The tests for gmtime and mktime functions pass. Some tests for strftime would fail (because not all formats are implemented, or because padding is not implemented), so I set them to be ignored for now.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: ia32-generic-qemu, armv7a9-zynq7000-zturn

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

**Review by nalajcie — COMMENTED** [review]
Great that you've moved the obsolete tests to unity framework!

I've put some nitpicks found by reading the code (I didn't run it).

File: libc/time/gmtime.c [hunk 1]
```diff
@@ -16,7 +16,9 @@
 #include <time.h>
 #include <stdio.h>
 #include <stdlib.h>
-#include "test_common.h"
+#include "../../test_common.h"
```
[the remark this answers is not in the document]
Comment (jmaksymowicz) [reply libc/time/gmtime.c]: OK, I made a new header file with common functions for time tests.

File: libc/time/gmtime.c [hunk 2]
```diff
@@ -25,77 +27,43 @@ static const time_t input_vector[T1_LEN];
...
-			tm_to_str(&t_exp, buff);
-			printf("Expected: %s \n\n", buff);
-		}
-		return 1;
-	}
-	else
-	{
-		return 0;
-	}
+	TEST_ASSERT_FALSE_MESSAGE(compare_tm(t, &t_exp), "Incorrect struct tm");
```
[the remark this answers is not in the document]
Comment (jmaksymowicz) [reply libc/time/gmtime.c]: Done.

File: libc/time/strftime.c [hunk 3]
```diff
@@ -69,47 +79,84 @@ static const struct test_data test_vector[T1_LEN] = {
...
 }
 
 
-int main(void)
+TEST(time_strftime, additional_format_chars)
 {
-	printf("STRFTIME TEST STARTED\n");
-	save_env();
+#ifdef __phoenix__
+	TEST_IGNORE();
```
[the remark this answers is not in the document]
Comment (jmaksymowicz) [reply libc/time/strftime.c]: I added messages to tests that are ignored for now. The issue is https://github.com/phoenix-rtos/phoenix-rtos-project/issues/351.

File: libc/time/host_generation.c [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	struct tm t;
+	srand(time(NULL));
+	for (i = 0; i < input_length; i++) {
+		t = (struct tm) {
+			.tm_sec = rand() % 59,
+			.tm_min = rand() % 59,
+			.tm_hour = rand() % 23,
+			.tm_mday = rand() % 30,
+			.tm_mon = rand() % 11,
+			.tm_year = 1980 + rand() % 60,
```
Comment (damianloew) [inline libc/time/host_generation.c:82]: There is an issue mentioned above.
```suggestion
			.tm_year = 80 + rand() % 60,
```

File: libc/time/strftime.c [hunk 5]
```diff
@@ -31,28 +31,48 @@ struct test_data {
...
-
-static const struct tm t3 = { .tm_sec = 11, .tm_min = 12, .tm_hour = 13, .tm_mday = 23, .tm_mon = 11,
-							  .tm_year = 105, .tm_wday = 1, .tm_yday = 235, .tm_isdst = 0 };
+static const struct tm t1 = { .tm_sec = 1, .tm_min = 1, .tm_hour = 6, .tm_mday = 2, .tm_mon = 2, .tm_year = 111, .tm_wday = 0, .tm_yday = 2, .tm_isdst = 0 };
+static const struct tm t2 = { .tm_sec = 11, .tm_min = 12, .tm_hour = 13, .tm_mday = 23, .tm_mon = 11, .tm_year = 95, .tm_wday = 6, .tm_yday = 235, .tm_isdst = 0 };
+static const struct tm t3 = { .tm_sec = 11, .tm_min = 12, .tm_hour = 13, .tm_mday = 23, .tm_mon = 11, .tm_year = 105, .tm_wday = 1, .tm_yday = 235, .tm_isdst = 0 };
+static const struct tm t4 = { .tm_sec = 0, .tm_min = 0, .tm_hour = 0, .tm_mday = 2, .tm_mon = 0, .tm_year = 99, .tm_wday = 6, .tm_yday = 2, .tm_isdst = 0 };
 
-static const struct test_data test_vector[T1_LEN] = {
+static const struct test_data basic_formatting[] = {
```
Comment (damianloew) [inline libc/time/strftime.c:39]: I think that we could place the declarations at the beginning of the file
like it's done in the rest of time tests. So sth like:
```suggestion
static const struct test_data basic_formatting[T1_LEN];
static const struct test_data additional_format_chars[T2_LEN];
...

static const struct test_data basic_formatting[] = {
```

Comment (damianloew) [reply libc/time/strftime.c:39]: Then we could place test data below `TEST_GROUP_RUNNER`.

Comment (damianloew) [reply libc/time/strftime.c:39]: Alternatively you could switch to the convention applied in this test in mktime/strftime tests - we just want to be consistent.

File: libc/time/gmtime.c [hunk 6]
```diff
@@ -25,77 +26,43 @@ static const time_t input_vector[T1_LEN];
...
+
+TEST_GROUP_RUNNER(time_gmtime)
+{
+	RUN_TEST_CASE(time_gmtime, basic_test);
+}
+
+
 static const time_t input_vector[] = {
 	/* 29 FEB 2016 - leap year */
 	61414866000,
```
[the remark this answers is not in the document]
Comment (damianloew) [reply libc/time/gmtime.c:68]: Same for mktime tests.

**Review by damianloew — COMMENTED** [review]
Thanks for applying the suggestions, please rebase changes and we can merge it :)

**Review by damianloew — APPROVED** [review]
LGTM

**Pull request merged**